### PR TITLE
Keys for the table data

### PR DESCRIPTION
**Pull request by vivekmenezes**
Keys for the table data

### CODE REVIEW

File: keys/constants.go [hunk 1]
```diff
@@ -117,6 +117,11 @@ var (
 	// LocalMax is the end of the local key range.
 	LocalMax = LocalPrefix.PrefixEnd()
 
+	// TableDataPrefix prefixes all Table data to aid in transitioning
+	// key:value data to Table data, and for ease of debugging.
+	TableDataPrefix       = proto.Key("table-")
+	TableDataPrefixLength = 6
```
Comment (petermattis) [inline keys/constants.go]: `s/6/len(TableDataPrefix)/g`?

File: keys/keys.go [hunk 2]
```diff
@@ -70,6 +70,42 @@ func MakeTableMetadataKey(namespaceID uint32, tableName string) proto.Key {
...
+	key = encoding.EncodeUvarint(key, uint64(tableID))
+	key = encoding.EncodeUvarint(key, uint64(indexID))
+	for _, column := range columns {
+		key = append(key, column...)
+	}
+	return key
+}
+
+// MakeTableIndexKey returns a primary or a secondary index key.
+func MakeTableIndexKey(tableID, indexID uint32, columns ...[]byte) proto.Key {
```
Comment (petermattis) [inline keys/keys.go]: Is `columns` intended to be column names are column values? If the latter, the name of the variable is confusing.

File: keys/keys.go [hunk 3]
```diff
@@ -70,6 +70,42 @@ func MakeTableMetadataKey(namespaceID uint32, tableName string) proto.Key {
...
+}
+
+// populateTableIndexKey populates the key passed in with the
+// order encoded values forming the index key.
+func populateTableIndexKey(key []byte, tableID, indexID uint32, columns ...[]byte) []byte {
+	key = append(key, TableDataPrefix...)
+	key = encoding.EncodeUvarint(key, uint64(tableID))
+	key = encoding.EncodeUvarint(key, uint64(indexID))
+	for _, column := range columns {
+		key = append(key, column...)
```
Comment (petermattis) [inline keys/keys.go]: You need to use `EncodeBytes` on each of the columns. Just appending them as you've done is incorrect. Consider 2 rows, one with the columns `"a", "ab"` and the other with the columns `"aa", "a"`. Just appending the columns together you get `"aab"` for one key and `"aaa"` for other key and then these sort incorrectly. Using `EncodeBytes` we would get the keys `"a\x00\x01ab\x00\x01"` and `"aa\x00\x01a\x00\x01"` which sort correctly.

The encoding routines have tests for all of this, but you need to add some simple tests for the ordering of the index and data keys.

File: keys/constants.go (hunk 1 above)
Comment (petermattis) [inline keys/constants.go]: Seeing how this value is used, I'd just nuke it and use `len(TableDataPrefix)` when needed.

File: keys/keys.go [hunk 4]
```diff
@@ -70,6 +70,42 @@ func MakeTableMetadataKey(namespaceID uint32, tableName string) proto.Key {
...
+}
+
+// MakeTableDataKey returns a key to a value at a specific row and column
+// in the table.
+func MakeTableDataKey(tableID, indexID, columnID uint32, columns ...[]byte) proto.Key {
+	k := make([]byte, 0, TableDataPrefixLength+30+indexKeyWidth(columns...))
+	k = populateTableIndexKey(k, tableID, indexID, columns...)
+	k = encoding.EncodeUvarint(k, uint64(columnID))
+	return k
+}
```
Comment (petermattis) [inline keys/keys.go]: I'm pretty sure we'll need decoding routines as well. See my `pmattis/client-table-impl` branch and `ScanStruct` for an example of where they will be necessary.

Comment (vivekmenezes) [inline keys/keys.go]: Let me add those in a separate PR

File: keys/keys.go (hunk 3 above)
Comment (vivekmenezes) [inline keys/keys.go]: Good point! Thanks

File: keys/keys.go [hunk 5]
```diff
@@ -70,6 +70,42 @@ func MakeTableMetadataKey(namespaceID uint32, tableName string) proto.Key {
...
+	key = encoding.EncodeUvarint(key, uint64(indexID))
+	for _, value := range columnValues {
+		key = encoding.EncodeBytes(key, value)
+	}
+	return key
+}
+
+// MakeTableIndexKey returns a primary or a secondary index key.
+func MakeTableIndexKey(tableID, indexID uint32, columnValues ...[]byte) proto.Key {
+	k := make([]byte, 0, len(TableDataPrefix)+20+indexKeyWidth(columnValues...))
```
Comment (tamird) [inline keys/keys.go]: what is this magic number?

File: keys/keys.go [hunk 6]
```diff
@@ -70,6 +70,42 @@ func MakeTableMetadataKey(namespaceID uint32, tableName string) proto.Key {
...
+	k = populateTableIndexKey(k, tableID, indexID, columnValues...)
+	return k
+}
+
+// MakeTableDataKey returns a key to a value at a specific row and column
+// in the table.
+func MakeTableDataKey(tableID, indexID, columnID uint32, columnValues ...[]byte) proto.Key {
+	k := make([]byte, 0, len(TableDataPrefix)+30+indexKeyWidth(columnValues...))
```
Comment (tamird) [inline keys/keys.go]: and this one

Comment (vivekmenezes) [inline keys/keys.go]: enough space to hold the encoded ints

Comment (tamird) [inline keys/keys.go]: please add a comment

File: keys/keys_test.go [hunk 7]
```diff
@@ -65,6 +66,19 @@ func TestKeyAddress(t *testing.T) {
...
+		// Ensure that all parameters in index/data keys are order encoded correctly.
+		{MakeTableIndexKey(12, 345, []byte("spencer")),
+			MakeKey(TableDataPrefix, encoding.EncodeUvarint(nil, 12), encoding.EncodeUvarint(nil, 345), encoding.EncodeBytes(nil, []byte("spencer")))},
+		{MakeTableDataKey(12, 345, 6, []byte("spencer")),
+			MakeKey(TableDataPrefix, encoding.EncodeUvarint(nil, 12), encoding.EncodeUvarint(nil, 345), encoding.EncodeBytes(nil, []byte("spencer")), encoding.EncodeUvarint(nil, 6))},
+		// Table data key tests with multiple columns encoded.
+		{MakeTableIndexKey(123, 234, []byte("a"), []byte("ba")), proto.Key("table-\t{\t\xeaa\x00\x01ba\x00\x01")},
+		{MakeTableIndexKey(123, 234, []byte("aa"), []byte("a")), proto.Key("table-\t{\t\xeaaa\x00\x01a\x00\x01")},
+		{MakeTableDataKey(123, 0, 12, []byte("a"), []byte("ba")), proto.Key("table-\t{\ba\x00\x01ba\x00\x01\t\f")},
+		{MakeTableDataKey(123, 0, 12, []byte("aa"), []byte("a")), proto.Key("table-\t{\baa\x00\x01a\x00\x01\t\f")},
```
Comment (petermattis) [inline keys/keys_test.go]: This isn't the correct place to be adding these test cases. `TestKeyAddress` is testing the `KeyAddress()` method. You really want new `TestMakeTableDataKey` and `TestMakeTableIndexKey` tests. And the tests should verify that ordering works correctly (which this test isn't doing).

File: keys/constants.go [hunk 8]
```diff
@@ -117,6 +117,10 @@ var (
 	// LocalMax is the end of the local key range.
 	LocalMax = LocalPrefix.PrefixEnd()
 
+	// TableDataPrefix prefixes all Table data to aid in transitioning
+	// key:value data to Table data, and for ease of debugging.
+	TableDataPrefix = proto.Key("table-")
```
Comment (petermattis) [inline keys/constants.go]: I'd place this key definition down lower in the list of constants. It feels out of place here wedged between local keys and system keys. Down at the very bottom of this `var` block seems right.

File: keys/keys.go [hunk 9]
```diff
@@ -70,6 +70,54 @@ func MakeTableMetadataKey(namespaceID uint32, tableName string) proto.Key {
...
+// indexID, and column values.
+//
+// This cap is inaccurate because the size of the encoding varies, depending
+// on the ints and the bytes being encoded. We really don't care, as long as
+// it provides a cap that prevents the append() builtin, used to populate the
+// buffer, from frequently reallocating more space.
+func indexKeyMaxBufferWidth(columnValues ...[]byte) (width int) {
+	// Accomodate the constant prefix, tableID, and indexID. 10 bytes
+	// for each ID.
+	width += len(TableDataPrefix) + 20
```
Comment (petermattis) [inline keys/keys.go]: `s/20/maxVarintSize/g` (you'll need to rebase to master to get `maxVarintSize`).

File: keys/keys.go [hunk 10]
```diff
@@ -70,6 +70,54 @@ func MakeTableMetadataKey(namespaceID uint32, tableName string) proto.Key {
...
+func MakeTableIndexKey(tableID, indexID uint32, columnValues ...[]byte) proto.Key {
+	k := make([]byte, 0, indexKeyMaxBufferWidth(columnValues...))
+	k = populateTableIndexKey(k, tableID, indexID, columnValues...)
+	return k
+}
+
+// MakeTableDataKey returns a key to a value at a specific row and column
+// in the table.
+func MakeTableDataKey(tableID, indexID, columnID uint32, columnValues ...[]byte) proto.Key {
+	// 10 bytes for the columnID
```
Comment (petermattis) [inline keys/keys.go]: Delete the comment.

File: keys/keys.go [hunk 11]
```diff
@@ -70,6 +70,54 @@ func MakeTableMetadataKey(namespaceID uint32, tableName string) proto.Key {
...
+	k := make([]byte, 0, indexKeyMaxBufferWidth(columnValues...)+10)
```
Comment (petermattis) [inline keys/keys.go]: `s/10/maxVarintSize/g`

File: keys/keys_test.go [hunk 12]
```diff
@@ -75,6 +76,25 @@ func TestKeyAddress(t *testing.T) {
 	}
 }
 
+func TestMakeTableIndexKey(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	key := MakeTableIndexKey(12, 345, []byte("foo"), []byte("bar"))
+	expKey := MakeKey(TableDataPrefix, encoding.EncodeUvarint(nil, 12), encoding.EncodeUvarint(nil, 345), encoding.EncodeBytes(nil, []byte("foo")), encoding.EncodeBytes(nil, []byte("bar")))
```
Comment (petermattis) [inline keys/keys_test.go]: This is just testing that `MakeTableIndexKey` has a particular implementation. What I'd rather see is something which generated a few index keys using different table, index and column values and verified they sort correctly. For example, something like:

```
  keys := []proto.Key{
    MakeTableIndexKey(0, 0, []byte("foo")),
    MakeTableIndexKey(0, 0, []byte("fooo")),
    MakeTableIndexKey(0, 1, []byte("bar")),
    MakeTableIndexKey(1, 0, []byte("bar")),
    MakeTableIndexKey(1, 0, []byte("bar"), []byte("foo")),
  }
  for i := 1; i < len(keys); i++ {
    if keys[i-1] >= keys[i] {
      t.Errorf("key %d >= key %d", i - 1, i)
    }
  }
```

**Comment by petermattis** [comment]
@vivekmenezes Github is far inferior to Mondrian when it comes to code reviews. I do not receive notification when you've uploaded a new patch set and cannot easily tell when you've addressed a comment. It is good practice to respond to every comment someone makes on your PR with a "Done" or "Note done and here is why...".

File: keys/constants.go (hunk 8 above)
Comment (vivekmenezes) [inline keys/constants.go]: Done

File: keys/keys.go (hunk 9 above)
Comment (vivekmenezes) [inline keys/keys.go]: Changed to binary.MaxVarintLen64

File: keys/keys.go (hunk 10 above)
Comment (vivekmenezes) [inline keys/keys.go]: done

File: keys/keys.go (hunk 11 above)
Comment (vivekmenezes) [inline keys/keys.go]: same

File: keys/keys_test.go (hunk 12 above)
Comment (vivekmenezes) [inline keys/keys_test.go]: Added in more unittest coverage and pretty much stole what you recommended. Thanks

**Comment by vivekmenezes** [comment]
Ready for another viewing

File: keys/keys.go [hunk 13]
```diff
@@ -62,7 +59,7 @@ func NodeStatusKey(nodeID int32) proto.Key {
 
 // MakeNameMetadataKey returns the key for the namespace.
 func MakeNameMetadataKey(parentID uint32, name string) proto.Key {
-	k := make([]byte, 0, len(NameMetadataPrefix)+maxVarintSize+len(name))
+	k := make([]byte, 0, len(NameMetadataPrefix)+binary.MaxVarintLen64+len(name))
```
Comment (petermattis) [inline keys/keys.go]: We're not using `encoding/binary` for encoding our varints, so it is incorrect to use `binary.MaxVarintLen64` here.

**Comment by petermattis** [comment]
LGTM modulo the `binary.MaxVarintLen64` issue.

File: keys/keys.go (hunk 13 above)
Comment (vivekmenezes) [inline keys/keys.go]: I see it being used by Encoding() in 

util/encoding/encoding.go

Should I rename maxVarintSize to MaxEncodeUvarint and export it from util/encoding/encoding.go package then?

Comment (tamird) [inline keys/keys.go]: > Should I rename maxVarintSize to MaxEncodeUvarint and export it from util/encoding/encoding.go package then?

Yep

Comment (petermattis) [inline keys/keys.go]: The use of `binary.PutVarint` in `util/encoding.Encode` should really be removed. Looks like the only use of that method is by `engine.Increment()` which appears to be unused.

Comment (vivekmenezes) [inline keys/keys.go]: Will look at taking that out in a new PR

File: util/encoding/encoding.go [hunk 14]
```diff
@@ -287,6 +287,11 @@ func DecodeVarintDecreasing(b []byte) ([]byte, int64) {
 	return leftover, ^v
 }
 
+const (
+	// MaxUvarintSize is the maximum size in bytes of an encoded uvarint.
+	MaxUvarintSize = 10
```
Comment (petermattis) [inline util/encoding/encoding.go]: `s/10/9/g`

**Comment by petermattis** [comment]
LGTM

File: util/encoding/encoding.go (hunk 14 above)
Comment (vivekmenezes) [inline util/encoding/encoding.go]: ah yes

File: keys/keys.go [hunk 15]
```diff
@@ -62,7 +58,7 @@ func NodeStatusKey(nodeID int32) proto.Key {
 
 // MakeNameMetadataKey returns the key for the namespace.
 func MakeNameMetadataKey(parentID uint32, name string) proto.Key {
-	k := make([]byte, 0, len(NameMetadataPrefix)+maxVarintSize+len(name))
+	k := make([]byte, 0, len(NameMetadataPrefix)+encoding.MaxUvarintSize+len(name))
```
Comment (tamird) [inline keys/keys.go]: this is still incorrect

Comment (tamird) [inline keys/keys.go]: disregard, `encoding` is util/encoding!

**Pull request merged**